src/core/mempool.rs

```diff
@@ -283,6 +285,130 @@ impl MemPoolDB {
+        }
+
+        let mut next_tips = MemPoolDB::get_chain_tips_at_height(&self.db, next_height)?;
+        if next_tips.len() == 0 {
+            // we're done -- no more chain tips
+            debug!("Done scanning mempool -- no chain tips at height {}", next_height);
+            return Ok(None);
+        }
+        
+        let ancestor_tip = {
+            let mut headers_tx = chainstate.headers_tx_begin()?;
+            // `next_height` is 1-indexed, since stacks blocks start at height 1.
+            // The headers DB, however, is 0-indexed, since it includes the genesis
+            // block header hash.  Account for this.
+            let next_header_height = next_height.saturating_sub(1);
+            let cur_index_hash = StacksBlockHeader::make_index_block_hash(tip_burn_header_hash, tip_block_hash);
+            match StacksChainState::get_index_tip_ancestor(&mut headers_tx, &cur_index_hash, next_header_height)? {
+                Some(tip_info) => tip_info,
+                None => {
+                    // no such ancestor.  We're done
+                    debug!("Done scanning mempool -- no ancestor at height {} off of {}/{} ({})", next_height, tip_burn_header_hash, tip_block_hash, StacksBlockHeader::make_index_block_hash(tip_burn_header_hash, tip_block_hash));
+                    return Ok(None);
+                }
+            }
+        };
+        
+        // find out which tip is the ancestor tip
+        let mut found = false;
+        let mut next_tip_burn_header_hash = tip_burn_header_hash.clone();
+        let mut next_tip_block_hash = tip_block_hash.clone();
+
+        for (burn_bhh, block_bhh) in next_tips.drain(..) {
+            if ancestor_tip.burn_header_hash == burn_bhh && ancestor_tip.anchored_header.block_hash() == block_bhh {
+                found = true;
+                next_tip_burn_header_hash = burn_bhh;
+                next_tip_block_hash = block_bhh;
+                break;
+            }
+        }
+        
+        if !found {
+            // no such ancestor.  We're done.
+            debug!("Done scanning mempool -- none of the available prior chain tips at {} is an ancestor of {}/{}", next_height, tip_burn_header_hash, tip_block_hash);
+            return Ok(None);
+        }
+
+        let next_timestamp = match MemPoolDB::get_next_timestamp(&self.db, &next_tip_burn_header_hash, &next_tip_block_hash, 0)? {
+            Some(ts) => ts,
+            None => {
+                unreachable!("No transactions at a chain tip that exists");
+            }
+        };
+        
+        debug!("Will start scaning mempool at {}/{} height={} ts={}", &next_tip_burn_header_hash, &next_tip_block_hash, next_height, next_timestamp);
+        Ok(Some((next_tip_burn_header_hash, next_tip_block_hash, next_height, next_timestamp)))
+    }
+
+    ///
+    /// Iterate over candidates in the mempool
+    ///  todo may be called multiple times, as the mempool walks back from
+    ///   the provided block.
```

Comment:
Might want to add a comment here that `todo` will be called once for each bundle of transactions at each ancestor chain tip from the given one, starting with the most recent chain tip and working backwards until there are no more transactions to consider.  Each batch of transactions passed to `todo` will be sorted in fee order, starting with the highest.

Reply:
Added in d0de5c9cc